File: 238-product-of-array-except-self/238-product-of-array-except-self.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        
        int[] right = new int[n];
        right[n-1] = 1;
        for (int i = n-2; i >= 0; i--) {
            right[i] = nums[i+1]*right[i+1];
        }
        int[] res = new int[n];
        res[0] = right[0];
        int left = 1;
        for (int i = 1; i < n; i++) {
            left *= nums[i-1];
            res[i]=left*right[i];
        }
        return res;
    }
}

File: 238-product-of-array-except-self/238-product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        left = [1] * n
        
        right = [1] * n
        
        for i in range(1, n):
            left[i] = nums[i-1]*left[i-1]
            right[n-i-1] = nums[n-i]*right[n-i]
            
        res = [1] * n
        for i in range(n):
            res[i] = left[i]*right[i]
            
        return res